modules/generator/registry/counter.go

```diff
@@ -126,7 +119,7 @@ func (c *counter) collectMetrics(appender storage.Appender, timeMs int64, extern
 
 	activeSeries = len(c.series)
 
-	lbls := make(labels.Labels, 1+len(externalLabels)+len(c.labels))
+	lbls := make(labels.Labels, 1+len(externalLabels)+4)
```

Comment:
Where does the 4 come from here?  Might be worth a comment on the code.

Reply:
good eye! I removed the line but @kvrhdn maybe has more context on whether that line is needed.

Reply:
In the original code we are preallocating the slice used by `labels.Builder` to ensure `lb.Set` doesn't have to expand the underlying array.

https://github.com/grafana/tempo/blob/089772bba3c50293ef3ac3fa615e7e1ff8a9ad73/modules/generator/registry/counter.go#L129-L130

I'm guessing the 4 must be something you set during testing. The sum should be:
- `1`: `__name__` (the metric name, always present)
- `len(externalLabels)`: constant labels injected by the registry config
- `len(c.labels)`: labels present on the counter itself

Since every serie can now have a different number of labels, it will be difficult to figure out that last value. You could calculate which series has the most labels, but iterating through all series is not going to be worth it.

I see the current code doesn't preallocate a slice anymore (it does `labels.NewBuilder(nil)`). This will still work but it might we might reallocate the slice several times.

Instead I'd suggest allocating at least enough space for the metric name and `externalLabels`, this data is easy to get and it's an easy win. As we are iterating through the series the labels slice might have to reallocated when it stumbles on a serie with more labels but that should only happen a couple of time.